File: R/aspect.R
#' Calculate topographic aspect from a DEM
#'
#' Calculates topographic aspect given a Digital Elevation Model (DEM) raster. Input and output are rasters of class \code{stars}, single-band (i.e., only `"x"` and `"y"` dimensions), with one attribute.
#'
#' @param x A raster (class \code{stars}) with two dimensions: \code{x} and \code{y}, i.e., a single-band raster, representing a DEM.
#' @param na_flag Value used to mark \code{NA} values in C code. This should be set to a value which is guaranteed to be absent from the input raster \code{x} (default is \code{-9999}).
#' @return A \code{stars} raster with topographic slope, i.e., the azimuth where the terrain is tilted towards, in decimal degrees (0-360) clockwise from north. Aspect of flat terrain, i.e., where all values in the neighborhood are equal, is set to \code{-1}. Returned raster values are of class \code{units} (decimal degrees).
#'
#' @note Aspect calculation results in \code{NA} when at least one of the cell neighbors is \code{NA}, including the outermost rows and columns. Given that the focal window size in aspect calculation is 3*3, this means that the outermost one row and one column are given an aspect value of \code{NA}.
#'
#' @references The topographic aspect algorithm is based on the \emph{How aspect works} article in the ArcGIS documentation:
#'
#' \url{https://desktop.arcgis.com/en/arcmap/10.3/tools/spatial-analyst-toolbox/how-aspect-works.htm}
#'
#' @examples
#' # Small example
#' data(dem)
#' dem_aspect = aspect(dem)
#' plot(
#'   dem, text_values = TRUE, breaks = "equal", 
#'   col = hcl.colors(11, "Spectral"), main = "input (elevation)"
#' )
#' plot(
#'   dem_aspect, text_values = TRUE, breaks = "equal", 
#'   col = hcl.colors(11, "Spectral"), main = "output (aspect)"
#' )
#' \donttest{
#' # Larger example
#' data(carmel)
#' carmel_aspect = aspect(carmel)
#' plot(
#'   carmel, breaks = "equal", 
#'   col = hcl.colors(11, "Spectral"), main = "input (elevation)"
#' )
#' plot(
#'   carmel_aspect, breaks = "equal", 
#'   col = hcl.colors(11, "Spectral"), main = "output (aspect)"
#' )
#' }
#'
#' @export

# Apply focal filter
aspect = function(x, na_flag = -9999) {

  # Checks
  x = check_one_attribute(x)
  x = check_2d(x)

  # Make template
  template = x

  # Number of extra lines
  steps = 1

  # To matrix
  input = layer_to_matrix(template, check = FALSE)
  input = matrix_extend(input, n = steps)

  # Matrix dimensions
  nrows = nrow(input)
  ncols = ncol(input)

  # Index matrix for the kernel
  kindex = NULL
  for(n in -steps:steps)    # Index for columns
    for(m in -steps:steps)  # Index for rows
      kindex = c(kindex, n*ncols + m)

  # To vector
  input_vector = as.vector(t(input))

  # Replace 'NA' with 'na_flag'
  input_vector2 = input_vector
  input_vector2[is.na(input_vector2)] = na_flag
  output_vector2 = rep(na_flag, length(input_vector2))

  # Apply filter
  result = .C(
    "aspect_c",
    as.double(input_vector2),
    as.integer(nrows),
    as.integer(ncols),
    as.integer(kindex),
    as.double(na_flag),
    as.double(output_vector2)
  )
  output_vector = result[[length(result)]]

  # Back to 'stars'
  output_vector[output_vector == na_flag] = NA
  output = matrix(output_vector, nrow(input), ncol(input), byrow = TRUE)
  output = matrix_trim(output, n = steps)
  output = t(output)
  template[[1]] = output

  # Set name and units
  names(template) = "aspect"
  template[[1]] = units::set_units(template[[1]], "degree")


  # Return
  return(template)

}

File: src/aspect.c
#define R_NO_REMAP
#include <math.h>

void aspect_c(
    double *input, 
    int *nrows, 
    int *ncols, 
    int *kindex, 
    double *na_flag, 
    double *output
    ) {
    
    int i, j, q;           // Counters
    long index;            // Index of the cell to be processed
    double current;        // Current value
    double dat[9];         // Data for current neighborhood
    int any_na;            // Flag for 'NA' in neighborhood
    double dzdx, dzdy, a;  // For aspect calculation

    // Loop over columns
    for(i = 1; i < *ncols - 1; i++) {

        // Loop over rows
        for(j = 1; j < *nrows - 1; j++) {

            for(q = 0; q < 9; q++) { dat[q] = *na_flag; }
            index = j * *ncols + i;
            any_na = 0;

            // Loop over weights matrix
            for(q = 0; q < 9; q++) {

                current = input[index + kindex[q]];

                // If current value is 'NA'
                if(current == *na_flag) {
                    any_na = 1;
                    continue;
                }

                // If current value is not 'NA'
                if(current != *na_flag) {
                    dat[q] = current;
                }

            }

            // If all values are valid, calculate aspect
            if(any_na == 0) {

                // Check flatness
                int equal = 0;
                for(q = 0; q < 8; q++) { 
                    if(dat[q] == dat[q+1]) { equal++; } 
                }

                // Flat terrain
                if(equal == 8) { 

                    output[index] = -1; 
                
                } else {

                    // Non-flat terrain
                    dzdx = ((dat[2] + 2*dat[5] + dat[8]) - (dat[0] + 2*dat[3] + dat[6])) / 8;
                    dzdy = ((dat[6] + 2*dat[7] + dat[8]) - (dat[0] + 2*dat[1] + dat[2])) / 8;
                    a = (180 / M_PI) * atan2(dzdy, -dzdx);
                    if (a < 0) { 
                        a = 90 - a;
                    } else {
                        if (a > 90) {
                            a = 360 - a + 90;
                        } else {
                            a = 90 - a;
                        }
                    }
                    output[index] = a;

                }

            }

            // Any 'NA' -> set 'NA'
            if(any_na == 1) {
                output[index] = *na_flag;
            }
            
        }

    }

}

